---
title: "Gene Set Variation Analysis"
author: "Jack Gisby"
date: "`r Sys.Date()`"
output:
  github_document:
    toc: true
    toc_depth: 2
    fig_width: 10
    fig_height: 7
---

# Load data and code
## Load packages and custom functions

This chunk loads packages required for the notebook in addition to custom
functions in the `../R` directory. These include functions for loading data,
running differential expression and enrichment analyses and making plots. The
main functions and their arguments are described by comments within these files.

```{r load_functions, message=FALSE, warning=FALSE}

# load relevant packages
library(data.table)
library(SummarizedExperiment)
library(lmerTest)
library(GSVA)
library(biomaRt)
library(edgeR)

# set up plotting settings
library(ggplot2)
library(ggpubr)

theme_set(theme_pubr(border = TRUE))
theme_update(text = element_text(size = 8))

# load custom functions from the R/ directory
source("../R/preprocessing.R")
source("../R/de.R")

set.seed(1)

```

## Load RNA-seq data

The raw RNA-seq counts are loaded, transformed into `SummarizedExperiment` objects,
and split into the "Wave 1" and "Wave 2" cohorts. `SnowParam` is also initialised
for the differential expression analysis using the `variancePartition` package.

```{r load_rna_data, message=FALSE, warning=FALSE, fig.show='hide'}

# get the raw RNA-seq counts
htseq_counts <- data.frame(fread("../data/htseq_counts.csv"))

mart <- useDataset("hsapiens_gene_ensembl", useMart("ensembl", host="uswest.ensembl.org"))

# get the wave 1 and wave 2 counts and their associated metadata
wave1_counts <- get_summarized_experiment(
  htseq_counts, 
  "../data/w1_metadata.csv", 
  mart = mart
)

wave2_counts <- get_summarized_experiment(
  htseq_counts, 
  "../data/w2_metadata.csv", 
  mart = mart
)

```

## Load SOMA protein data

The SOMA data is loaded into wide format (a matrix of samples x proteins) and 
split into the "Wave 1" and "Wave 2" cohorts by the `get_soma_data` function. 

```{r load_prot_data, message=FALSE, warning=FALSE, fig.show='hide'}

wave1_soma <- get_soma_data(
  soma_abundance = "../../data/20210819_soma_processed/soma_abundances.csv", 
  sample_meta = "../../scripts/20210713_salmon_counts_initial_look/w1_metadata.csv", 
  sample_technical_meta = "../../data/20210819_soma_processed/sample_meta.csv", 
  feature_meta = "../../data/20210819_soma_processed/feature_meta.csv",
  ret_wide = TRUE
)

# convert column names from sequence IDs to gene IDs
colnames(wave1_soma$soma_abundance) <- sapply(colnames(wave1_soma$soma_abundance), function(cname) {
  wave1_soma$feature_meta$gene_id[wave1_soma$feature_meta$converted_seq_ids == cname]
}) 

# remove duplicated genes
wave1_soma$soma_abundance <- wave1_soma$soma_abundance[, !duplicated(colnames(wave1_soma$soma_abundance))]
wave1_soma$feature_meta <- wave1_soma$feature_meta[!duplicated(wave1_soma$feature_meta$converted_seq_ids) ,]

wave2_soma <- get_soma_data(
  soma_abundance = "../data/soma_abundance.csv", 
  sample_meta = "../data/w2_metadata.csv", 
  sample_technical_meta = "../data/sample_technical_meta.csv", 
  feature_meta = "../data/feature_meta.csv",
  ret_wide = TRUE
)

# convert column names from sequence IDs to gene IDs
colnames(wave2_soma$soma_abundance) <- sapply(colnames(wave2_soma$soma_abundance), function(cname) {
  wave2_soma$feature_meta$gene_id[wave2_soma$feature_meta$converted_seq_ids == cname]
}) 

# remove duplicated genes
wave2_soma$soma_abundance <- wave2_soma$soma_abundance[, !duplicated(colnames(wave2_soma$soma_abundance))]
wave2_soma$feature_meta <- wave2_soma$feature_meta[!duplicated(wave2_soma$feature_meta$converted_seq_ids) ,]

```

## Setup Gene Set Variation Analysis 

We use the MsigDB C2 gene sets to create a GeneSetCollection object for GSVA.

The symbols file has been edited to include two additional gene sets.

```{r generate_gsva_object, message=FALSE, warning=FALSE, eval=FALSE, include=TRUE}

# reads msigdb release 7.4 symbols, edited to include two additional modules
gsc.file <- scan("data/20210811_gsva/c2.cp.v7.4.symbols.edited.txt", what = "", sep="\n")
gsclist <- strsplit(gsc.file, "[[:space:]]+")
names(gsclist) <- sapply(gsclist, `[[`, 1)

gsclist <- lapply(gsclist, `[`, -1)
gsclist <- lapply(gsclist, `[`, -1)

uniqueList <- lapply(gsclist, unique)

makeSet <- function(geneIds, n) {GeneSet(geneIds, geneIdType=SymbolIdentifier(), setName=n)}
gsclist <- mapply(makeSet, uniqueList[], names(gsclist))

gsc <- GeneSetCollection(gsclist)

```

```{r load_gsva_object, eval=TRUE, include=FALSE}

# instead of reading in the GSVA modules in the notebook,
# just read a copy saved as an RDS
gsc <- readRDS("../results/2_gsva_analysis/terms.r")

```


# Differential expression analyses of COVID-19 status

The differential expression analyses performed in this document are essentially
identical to the analyses performed in the `1_differential_expression.Rmd` notebook,
however here we analyse gene sets rather than individual genes and proteins. 

We also do not compare pre-infection and recovery samples using GSVA. 

## RNA-seq analysis

In this section, we calculate the gene set x sample matrix for each cohort and
compare COVID-19 positive vs. negative samples for each cohort using linear
mixed models.

### Wave 1 transcriptomic COVID-19 infected vs. non-infected comparison

```{r wave1_case_control_gene_gsva, message=FALSE, warning=FALSE}

# apply normalisation to counts (returns edgeR DGE object)
wave1_case_control_se <- normalize_se(
  wave1_counts,
  normalisation_level = "logcpm",
  group = "case_control"
)

# remove duplicated genes
wave1_case_control_se <- wave1_case_control_se[!duplicated(rowData(wave1_case_control_se)$gene_id), ]
rownames(wave1_case_control_se) <- rowData(wave1_case_control_se)$gene_id

# calculate GSVA (gene set x samples matrix)
wave1_case_control_eset <- gsva(
  assay(wave1_case_control_se, 2), 
  gsc, 
  min.sz = 10, 
  max.sz = Inf, 
  parallel.sz = 8
)

# apply linear mixed models to gene sets
wave1_case_control_eset_models <- eset_de(
  se = wave1_case_control_se,
  eset = wave1_case_control_eset,
  formula_string = "set_expr ~ case_control + sex + ethnicity + calc_age + (1 | individual_id)"
)

# function to convert list of models to table of results
get_eset_results <- function(models, wave, data_type, comparison) {
    
    eset_de <- data.frame()
    
    if (comparison == "negative_vs_positive") {
        t <- "case_controlPOSITIVE"
        anova_col <- "case_control"
    } else {
        t <- "WHO_temp_severity.L"
        anova_col <- "WHO_temp_severity"
    }
    
    for (i in 1:length(models)) {
      set_anova <- anova(models[[i]])
      set_summ <- summary(models[[i]])
    
      eset_de <- rbind(eset_de, data.frame(
        term_id = names(models)[i],
        pval = set_anova$`Pr(>F)`[rownames(set_anova) == anova_col],
        beta = set_summ$coefficients[rownames(set_summ$coefficients) == t, colnames(set_summ$coefficients) == "Estimate"],
        se = set_summ$coefficients[rownames(set_summ$coefficients) == t, colnames(set_summ$coefficients) == "Std. Error"]
      ))
    }
    
    eset_de$log_p <- -log10(eset_de$pval)
    eset_de$adj_p <- p.adjust(eset_de$pval, method = "BH")
    
    write.csv(eset_de, paste0("../results/2_gsva_analysis/wave_", wave, "/", data_type, "_", comparison, ".csv"), row.names = FALSE)
    
    return(eset_de)
}

wave1_case_control_eset_de <- get_eset_results(wave1_case_control_eset_models, 1, "gene", "negative_vs_positive")

```

### Wave 2 transcriptomic COVID-19 infected vs. pre-infection comparison

```{r wave2_case_control_gene_gsva, message=FALSE, warning=FALSE}

# apply normalisation to counts (returns edgeR DGE object)
wave2_case_control_se <- normalize_se(
  wave2_counts[, wave2_counts$case_control != "RECOVERY"],
  normalisation_level = "logcpm",
  group = "case_control"
)

# remove duplicated genes
wave2_case_control_se <- wave2_case_control_se[!duplicated(rowData(wave2_case_control_se)$gene_id), ]
rownames(wave2_case_control_se) <- rowData(wave2_case_control_se)$gene_id

# calculate GSVA (gene set x samples matrix)
wave2_case_control_eset <- gsva(
  assay(wave2_case_control_se, 2), 
  gsc, 
  min.sz = 10, 
  max.sz = Inf, 
  parallel.sz = 8
)

# apply linear mixed models to gene sets
wave2_case_control_eset_models <- eset_de(
  se = wave2_case_control_se,
  eset = wave2_case_control_eset,
  formula_string = "set_expr ~ case_control + sex + ethnicity + calc_age + (1 | individual_id)"
)

wave2_case_control_eset_de <- get_eset_results(wave2_case_control_eset_models, 2, "gene", "negative_vs_positive")

```

## Proteomic analysis

As for the transcriptomic data, we use GSVA to summarise the proteomics data into
sets of proteins, then use linear mixed models to associate them with COVID-19
status.

### Wave 1 proteomic COVID-19 infected vs. non-infected comparison

```{r wave1_case_control_prot_gsva, message=FALSE, warning=FALSE}

# apply GSVA to protein matrix
wave1_case_control_pset <- gsva(t(wave1_soma$soma_abundance), gsc, min.sz = 10, max.sz = Inf, verbose = TRUE, parallel.sz = 8)

# apply linear mixed models to protein sets
wave1_case_control_pset_models <- pset_de(
  se = t(wave1_soma$soma_abundance),
  pset = wave1_case_control_pset,
  col_data = wave1_soma$sample_meta,
  formula_string = "set_expr ~ case_control + sex + ethnicity + calc_age + (1 | individual_id)"
)

wave1_case_control_pset_de <- get_eset_results(wave1_case_control_pset_models, 1, "protein", "negative_vs_positive")

```

### Wave 2 proteomic COVID-19 infected vs. non-infected comparison

```{r wave2_case_control_prot_gsva, message=FALSE, warning=FALSE}

# apply GSVA to protein matrix
wave2_case_control_pset <- gsva(t(wave2_soma$soma_abundance), gsc, min.sz = 10, max.sz = Inf, verbose = TRUE, parallel.sz = 8)

# apply linear mixed models to protein sets
wave2_case_control_pset_models <- pset_de(
  se = t(wave2_soma$soma_abundance),
  pset = wave2_case_control_pset,
  col_data = wave2_soma$sample_meta,
  formula_string = "set_expr ~ case_control + sex + ethnicity + calc_age + (1 | individual_id)"
)

wave1_case_control_pset_de <- get_eset_results(wave2_case_control_pset_models, 2, "protein", "negative_vs_positive")

```

# Differential expression analyses of COVID-19 severity at time of sampling

Like the analyses performed in the `1_differential_expression.Rmd` notebook, we 
analyse COVID-19 positive samples by ordinal contemporaneous severity. However,
in this case we analyse gene and proteins sets rather than individual genes and 
proteins.

## RNA-seq analysis

### Wave 1 transcriptomic COVID-19 severity analysis

```{r wave1_severity_gene_gsva, message=FALSE, warning=FALSE}

# convert contemporaneous severity to an ordered factor
wave1_counts$WHO_temp_severity <- ordered(wave1_counts$WHO_temp_severity, c("mild", "moderate", "severe", "critical"))

# get normalised expression matrix for COVID-19 positive samples
wave1_severity_se <- normalize_se(
  wave1_counts[, wave1_counts$case_control == "POSITIVE" & !is.na(wave1_counts$WHO_temp_severity)],
  normalisation_level = "logcpm",
  group = "WHO_temp_severity"
)

wave1_severity_se <- wave1_severity_se[!duplicated(rowData(wave1_severity_se)$gene_id), ]
rownames(wave1_severity_se) <- rowData(wave1_severity_se)$gene_id

# calculate GSVA to gene expression matrix
wave1_severity_eset <- gsva(assay(wave1_severity_se, 2), gsc, min.sz = 10, max.sz = Inf, verbose = TRUE, parallel.sz = 8)

# apply linear mixed models to gene sets
wave1_severity_eset_models <- eset_de(
  se = wave1_severity_se,
  eset = wave1_severity_eset,
  formula_string = "set_expr ~ WHO_temp_severity + sex + ethnicity + calc_age + (1 | individual_id)"
)

wave1_severity_eset_de <- get_eset_results(wave1_severity_eset_models, 1, "gene", "severity")

```

### Wave 2 transcriptomic COVID-19 severity analysis

```{r wave2_severity_gene_gsva, message=FALSE, warning=FALSE}

# convert contemporaneous severity to an ordered factor
wave2_counts$WHO_temp_severity <- ordered(wave2_counts$WHO_temp_severity, c("mild", "moderate", "severe", "critical"))

# get normalised expression matrix for COVID-19 positive samples
wave2_severity_se <- normalize_se(
  wave2_counts[, wave2_counts$case_control == "POSITIVE" & !is.na(wave2_counts$WHO_temp_severity)],
  normalisation_level = "logcpm",
  group = "WHO_temp_severity"
)

wave2_severity_se <- wave2_severity_se[!duplicated(rowData(wave2_severity_se)$gene_id), ]
rownames(wave2_severity_se) <- rowData(wave2_severity_se)$gene_id

# calculate GSVA to gene expression matrix
wave2_severity_eset <- gsva(assay(wave2_severity_se, 2), gsc, min.sz = 10, max.sz = Inf, verbose = TRUE, parallel.sz = 8)

# apply linear mixed models to gene sets
wave2_severity_eset_models <- eset_de(
  se = wave2_severity_se,
  eset = wave2_severity_eset,
  formula_string = "set_expr ~ WHO_temp_severity + sex + ethnicity + calc_age + (1 | individual_id)"
)

wave2_severity_eset_de <- get_eset_results(wave2_severity_eset_models, 2, "gene", "severity")

```

## Proteomic analysis

### Wave 1 proteomic COVID-19 severity analysis

```{r wave1_severity_prot_gsva, message=FALSE, warning=FALSE}

# remove negative samples from protein matrix
pos_ids <- wave1_soma$sample_meta$sample_id[wave1_soma$sample_meta$case_control == "POSITIVE"]

wave1_soma_sev <- wave1_soma
wave1_soma_sev$soma_abundance <- wave1_soma_sev$soma_abundance[rownames(wave1_soma_sev$soma_abundance) %in% pos_ids ,]
wave1_soma_sev$sample_meta <- wave1_soma_sev$sample_meta[wave1_soma_sev$sample_meta$sample_id %in% pos_ids ,]

# convert contemporaneous severity to an ordered factor
wave1_soma_sev$sample_meta$WHO_temp_severity <- ordered(wave1_soma_sev$sample_meta$WHO_temp_severity, c("mild", "moderate", "severe", "critical"))

# calculate GSVA to protein matrix
wave1_severity_pset <- gsva(t(wave1_soma_sev$soma_abundance), gsc, min.sz = 10, max.sz = Inf, verbose = TRUE, parallel.sz = 8)

# apply linear mixed models to protein sets
wave1_severity_pset_models <- pset_de(
  se = t(wave1_soma_sev$soma_abundance),
  col_data = wave1_soma_sev$sample_meta,
  pset = wave1_severity_pset,
  formula_string = "set_expr ~ WHO_temp_severity + sex + ethnicity + calc_age + (1 | individual_id)"
)

wave1_severity_pset_de <- get_eset_results(wave1_severity_pset_models, 1, "protein", "severity")

```

### Wave 2 proteomic COVID-19 severity analysis

```{r wave2_severity_prot_gsva, message=FALSE, warning=FALSE}

# remove negative samples from protein matrix
pos_ids <- wave2_soma$sample_meta$sample_id[wave2_soma$sample_meta$case_control == "POSITIVE"]

wave2_soma_sev <- wave2_soma
wave2_soma_sev$soma_abundance <- wave2_soma_sev$soma_abundance[rownames(wave2_soma_sev$soma_abundance) %in% pos_ids ,]
wave2_soma_sev$sample_meta <- wave2_soma_sev$sample_meta[wave2_soma_sev$sample_meta$sample_id %in% pos_ids ,]

# convert contemporaneous severity to an ordered factor
wave2_soma_sev$sample_meta$WHO_temp_severity <- ordered(wave2_soma_sev$sample_meta$WHO_temp_severity, c("mild", "moderate", "severe", "critical"))

# calculate GSVA to protein matrix
wave2_severity_pset <- gsva(t(wave2_soma_sev$soma_abundance), gsc, min.sz = 10, max.sz = Inf, verbose = TRUE, parallel.sz = 8)

# apply linear mixed models to protein sets
wave2_severity_pset_models <- pset_de(
  se = t(wave2_soma_sev$soma_abundance),
  col_data = wave2_soma_sev$sample_meta,
  pset = wave2_severity_pset,
  formula_string = "set_expr ~ WHO_temp_severity + sex + ethnicity + calc_age + (1 | individual_id)"
)

wave2_severity_pset_de <- get_eset_results(wave2_severity_pset_models, 2, "protein", "severity")

```

# Session info

Session information collected at the point this markdown document was knit.

```{r session_info}

sessionInfo()

```
